Feature: Option to retain old grading data

pranavrao145 opened the pull request:
## Proposed Changes
Closes #5212.

## Type of Change
*(Write an `X` or a brief description next to the type or types that best describe your changes.)*

| Type                                                                                    | Applies? |
|-----------------------------------------------------------------------------------------|----------|
| 🚨 *Breaking change* (fix or feature that would cause existing functionality to change) |          |
| ✨ *New feature* (non-breaking change that adds functionality)                          |     X     |
| 🐛 *Bug fix* (non-breaking change that fixes an issue)                                  |          |
| 🎨 *User interface change* (change to user interface; provide screenshots)              |          |
| ♻️ *Refactoring* (internal change to codebase, without changing functionality)          |          |
| 🚦 *Test update* (change that *only* adds or modifies tests)                            |          |
| 📦 *Dependency update* (change that updates a dependency)                               |          |
| 🔧 *Internal* (change that *only* affects developers or continuous integration)         |          |


## Checklist
*(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)*

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [X] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable. 
  - Associated PR: https://github.com/MarkUsProject/Wiki/pull/221

After opening your pull request:

- [x] I have updated the project Changelog (this is required for all changes).
- [X] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments
Within inline comments.

pranavrao145 commented on app/jobs/submissions_job.rb:
```diff
@@ -5,6 +5,101 @@ def add_warning_messages(messages)
+      filename = annotation.submission_file.filename
+      path = annotation.submission_file.path
+
+      new_submission_file = new_submission.submission_files.where(filename: filename, path: path).first
```
I wonder whether there is a better way to do this (rather than by filename and path).

pranavrao145 commented on spec/jobs/submissions_job_spec.rb:
```diff
@@ -185,6 +185,182 @@
+
+        it 'retains the correct mark values for each result' do
+          @new_results.zip(@original_results).each do |new_result, old_result|
+            expect(new_result.marks.map(&:mark)).to eq(old_result.marks.map(&:mark))
```
This is a bit flaky - 1% of the time the marks come out in different orders and causes the test to fail. Is sorting the arrays and comparing solely the content still a good test?

pranavrao145 commented on app/assets/javascripts/Components/repo_browser.jsx:
```diff
@@ -168,15 +191,26 @@ class ManualCollectionForm extends React.Component {
-          <button
-            type="submit"
-            name="commit"
-            onClick={event => {
```
I moved this out of the `onClick` logic and into the `onSubmit` logic for the form so I could test this properly, otherwise there were errors about a lack of implementation for `onSubmit`

pranavrao145 commented on app/jobs/submissions_job.rb:
```diff
@@ -5,6 +5,101 @@ def add_warning_messages(messages)
+    ]
+
+    # copy over data from old result to new result
+    result_data.each do |result_set|
```
Same copy logic for annotations, marks, extra marks.

david-yz-liu replied:
Overall, there are enough differences in behaviour that I don't think this loop is worth it.

1. Only existing _marks_ need to be destroyed (this is the only model with some pre-existing objects)
2. Only _annotations_ need to have a separate check for an existing submission file (which you've correctly noted below may trigger a case where an annotation _isn't_ copied over).
3. For _extra marks_, you should copy over all point-based marks (these are set manually by the grader), but _ignore_ all percentage-based marks (these are set by the late penalty, which is something that the instructor decides to apply, or not, for the newly-collected submission). Please make sure to document the latter case, as it's a bit subtle.

pranavrao145 replied:
This has also been fixed, and I've added docs both within the code and on the Wiki!

pranavrao145 commented on jest.config.js:
```diff
@@ -60,7 +60,9 @@ module.exports = {
   // A set of global variables that need to be available in all test environments
-  // globals: {},
+  globals: {
+    AUTH_TOKEN: "test_auth_token",
```
Need this for Jest to work, `ManualCollectionForm` has a constant in it that's normally defined on layout render but not in component-specific testing.

pranavrao145 commented on app/jobs/submissions_job.rb:
```diff
@@ -5,6 +5,101 @@ def add_warning_messages(messages)
+      end
+
+      annotation.update(submission_file_id: new_submission_file.id)
+      add_warning_messages(annotation.errors.full_messages) if annotation.errors.present?
```
The approach I've taken to error handling is just adding all error messages to the flash message using the function already provided in this file - not sure if this is the best way to handle this.

david-yz-liu reviewed (commented):
Hi @pranavrao145, great work! I just did a quick review of the back-end implementation and left some comments. Please work on those, and then I'll do another review in a few days.

pranavrao145 commented:
I'm just going to convert this back to a draft while I work on addressing comments just so it doesn't run tests on intermediate work, and I'll make it ready for review again once I'm finished!

pranavrao145 commented on app/jobs/submissions_job.rb:
```diff
@@ -5,6 +5,101 @@ def add_warning_messages(messages)
     Rails.logger.error msg
   end
 
+  def copy_old_grading_data(new_submission, grouping)
```
Hi @david-yz-liu, thanks for the feedback! Code is looking a lot cleaner now. One question, seems the flash message on error is already taken care of implicitly when the Submissions model raises an exception, and it displays a clean message in red. However, this stops progress on all submissions if any one of them has a problem.

Would you like me to explicitly rescue Submission class errors here and add it only as a warning-level alert (using `add_warning_messages`), or just leave it at the error-level alert with a stop in processing other submissions? Content-wise, messages on both are pretty much the same (neither give the group info), so I thought I'd verify what the intended behaviour is.

david-yz-liu commented on app/assets/javascripts/Components/Modals/collect_submissions_modal.jsx:
```diff
@@ -96,15 +102,45 @@ class CollectSubmissionsModal extends React.Component {
+                    />
+                  </label>
+                  &nbsp;
+                  {this.state.retain_existing_grading ? (
```
Keep the checkbox label the same for both checked and unchecked states. Instead, if the option is unchecked, show a warning div (same styling as warning flash messages) with the extra warning message. Since the div will have good styling already, you don't need to bold the warning message.

david-yz-liu commented on app/assets/javascripts/Components/__tests__/repo_browser_manual_collection_form.test.jsx:
```diff
@@ -0,0 +1,72 @@
+
+// workaround needed for using i18n in jest tests, see
+// https://github.com/fnando/i18n/issues/26#issuecomment-1235751777
+jest.mock("i18n-js", () => {
```
Ah yes, can you actually move these mocks into one of the Jest setup files (and remove them from existing places in the codebase)?

pranavrao145 replied:
As it turns out, apparently we don't need these at all! Seems that there was already a Jest setup in `jest_env_setup.js` which defines `I18n` for all Jest tests globally. When writing this and previous tests, I never considered it may be implicitly defined so I didn't try without explicitly defining it - should have looked in some other files first, sorry about that! I've gone ahead and removed this setup from all files that use it.

david-yz-liu commented on app/controllers/submissions_controller.rb:
```diff
@@ -170,9 +170,16 @@ def manually_collect_and_begin_grading
                          else
                            params[:apply_late_penalty]
                          end
+    retain_existing_grading = if params[:retain_existing_grading].nil?
```
Similar to below, here I suspect you can do a `params[:retain_existing_grading] == 'true'`. You can modify the `apply_late_penalty` setting above as well.

david-yz-liu reviewed (commented):
@pranavrao145 great work. I just left a few minor comments.

A wrinkle: after some discussion, I now understand that a common use case for this feature is that when an instructor receives a remark request, that's what triggers a new submission to be collected (i.e., the submission files need to be changed). Given this, I do think it's best to copy over both the original result and new remark result after all. Can you please modify the code to make this work (you might want to refactor again to create a `Result` instance method). Sorry about this!

david-yz-liu commented on config/locales/views/submissions/en.yml:
```diff
@@ -16,11 +16,14 @@ en:
       manual_collection: Manual Collection
-      override_existing_html: 'Recollect previously collected submissions? <strong>Warning: any previous grading work will be lost.</strong>'
-      overwrite_warning: Collecting and grading this revision will overwrite any previous collections/grading done for this group on this assignment.  Are you sure you want to do this?
+      override_existing: Recollect previously collected submissions?
```
This message should be updated now (since remark data is now also copied)

pranavrao145 replied:
Ah yes, I will actually make this a confirmation message rather than a warning now because we're pretty much copying everything.

david-yz-liu commented on app/assets/javascripts/Components/repo_browser.jsx:
```diff
@@ -168,15 +191,26 @@ class ManualCollectionForm extends React.Component {
+                this.setState({retainExistingGrading: e.target.checked});
+              }}
+            />
+            <label
```
Use `htmlFor` attribute to link this label to the corresponding `input`; make sure to do this in the modal as well.

FYI, you can test this in the UI by *clicking on the label itself*: when properly linked, clicking the label should act the same as checking the checkbox itself.

david-yz-liu commented on app/controllers/submissions_controller.rb:
```diff
@@ -167,14 +167,13 @@ def manually_collect_and_begin_grading
-                           params[:apply_late_penalty]
-                         end
+    apply_late_penalty = params[:apply_late_penalty] == 'true'
+    retain_existing_grading = params[:retain_existing_grading] == 'true'
```
I'm not sure if you manually tested this after making this change, but it seems like the parameter value is `"on"`, not `"true"`, which is a bit strange.

pranavrao145 replied:
Ah whoops, I did not manually test this in this iteration of changes - sorry about that! However, I recall now that I actually tried to make this change a couple of weeks ago when messing around with this, and I did run into this problem of it being `on` instead of `true`, which is why I left it the way it was. I will go ahead and revert this change for now - thanks for the catch.

david-yz-liu commented on app/models/submission.rb:
```diff
@@ -255,6 +255,52 @@ def make_remark_result
+      self.create_result # additional result for remark data
+
+      # if there's already a remark result created as well
+      if old_submission.has_remark?
```
This check is redundant with the `old_submission.remark_request.present?` check above (unless I'm missing something?)

pranavrao145 replied:
So this is a bit odd, even to me - I'm overall trying to handle retaining data from two cases here:

1. In the case where the student is writing the remark request, has saved it, but has not submitted it to the instructor yet, the `remark_request` attribute and associated timestamp is populated on the submission, but no remark _result_ has been created yet; so I just check the attribute manually.
2. Where the student has _submitted_ the remark request to the instructor, everything in case 1 happens, but also a new remark result _is_ created, so we have to run the copy on that. Since `has_remark?` basically checks for if that remark result exists or not, I used it here to verify this case.

One thing worth noting is that `has_remark?` has the following comments:

> Returns a remark result that may or may not have been submitted. If you want the submitted remark result then the submitted_remark association should be used instead.

However, unless I'm misinterpreting what it means for a remark to not be submitted yet, I don't think this is fully accurate. From what I observed, I'm not actually sure it's possible to have a remark result without submitting it. To me, the function `has_remark?` seems to do practically the same thing as `submitted_remark`, but in a different way. Specifically, they both return the remark result (which I'm assuming is always the latest result if there is more than one result on the submission); `submitted_remark` gets it using the `remark_request_submitted_at` timestamp on `Result` and `has_remark` returns the last result directly _contingent_ on there being 2 results minimum). Wasn't sure which one to use so I just went with `has_remark?`, but `submitted_remark` might be better for readability.

Do you have any guidance on the best way to handle both of the above cases (or if I am even interpreting this correctly)?

david-yz-liu replied:
Ah I see, yes that makes sense. In this case I believe you can do the following:

1. *Always* copy the `remark_request` and `remark_request_timestamp` attributes from the old submission to the new submission; this doesn't need to be guarded by a conditional.
2. Move the `self.create_result` to inside the `old_submission.has_remark?` check. You can also simplify a bit by first doing something like `old_remark_result = old_submission.remark_result` and then checking whether `old_remark_result.present?`

david-yz-liu commented on app/models/result.rb:
```diff
@@ -135,6 +135,43 @@ def self.get_total_extra_marks(result_ids, max_mark: nil, user_visibility: :ta_v
+      # we just skip adding this annotation to the  new result
+      annotation_filename = annotation.submission_file.filename
+      annotation_path = annotation.submission_file.path
+      new_submission_file = self.submission.submission_files.where(filename: annotation_filename,
```
the search is okay, but use `find_by` instead of `.where(...).first`

david-yz-liu commented on app/models/result.rb:
```diff
@@ -135,6 +135,43 @@ def self.get_total_extra_marks(result_ids, max_mark: nil, user_visibility: :ta_v
+
+    old_result.annotations.each do |annotation|
+      # annotations are associated with files; if a file for an annotation doesn't exist
+      # we just skip adding this annotation to the  new result
```
delete the extra space

david-yz-liu reviewed (commented):
@pranavrao145 great work! I just left a few last comments, but I think the end is in sight. 😄

david-yz-liu reviewed (approved):
@pranavrao145 great job getting this completed! 🎉
